File: src/lib/actions/chores.ts
"use server";

import { and, eq, gte, lt, sql } from "drizzle-orm";
import { revalidatePath } from "next/cache";
import { db } from "@/db";
import { choreCompletions, chores, users } from "@/db/schema";
import { requireActiveUser, requireAdmin } from "@/lib/auth/current-user";
import { serializeRecurrenceDays, type Recurrence } from "@/lib/chores/recurrence";
import type { FormState } from "./form-state";

function parseDateInput(value: string): Date | null {
  if (!/^\d{4}-\d{2}-\d{2}$/.test(value)) return null;
  const date = new Date(`${value}T00:00:00`);
  return Number.isNaN(date.getTime()) ? null : date;
}

type ChoreFieldsResult =
  | { error: string }
  | {
      title: string;
      description: string | null;
      points: number;
      assignedToUserId: string;
      recurrence: Recurrence;
      dueDate: Date;
      recurrenceDays: string | null;
    };

function readChoreFields(formData: FormData): ChoreFieldsResult {
  const title = String(formData.get("title") ?? "").trim();
  const description = String(formData.get("description") ?? "").trim();
  const pointsRaw = String(formData.get("points") ?? "0");
  const points = Number.parseInt(pointsRaw, 10);
  const assignedToUserId = String(formData.get("assignedToUserId") ?? "");
  const recurrence = String(formData.get("recurrence") ?? "none") as Recurrence;
  const dueDateRaw = String(formData.get("dueDate") ?? "");
  const recurrenceDayValues = formData.getAll("recurrenceDays").map(String);

  if (!title) return { error: "Title is required." };
  if (!assignedToUserId) return { error: "Choose who this is assigned to." };
  if (!Number.isFinite(points) || points < 0) {
    return { error: "Points must be zero or a positive number." };
  }

  const dueDate = parseDateInput(dueDateRaw);
  if (!dueDate) return { error: "Choose a valid due date." };

  if (
    !["none", "daily", "weekly", "weekdays", "custom"].includes(recurrence)
  ) {
    return { error: "Invalid recurrence." };
  }
  if (recurrence === "custom" && recurrenceDayValues.length === 0) {
    return { error: "Pick at least one day for a custom schedule." };
  }

  return {
    title,
    description: description || null,
    points,
    assignedToUserId,
    recurrence,
    dueDate,
    recurrenceDays:
      recurrence === "custom"
        ? serializeRecurrenceDays(recurrenceDayValues.map(Number))
        : null,
  };
}

export async function createChore(
  _prevState: FormState,
  formData: FormData,
): Promise<FormState> {
  const { household, user } = await requireActiveUser();
  requireAdmin(user);

  const fields = readChoreFields(formData);
  if ("error" in fields) return { error: fields.error };

  const assignee = await db.query.users.findFirst({
    where: and(
      eq(users.id, fields.assignedToUserId),
      eq(users.householdId, household.id),
    ),
  });
  if (!assignee) return { error: "Assignee not found." };

  await db.insert(chores).values({
    id: crypto.randomUUID(),
    householdId: household.id,
    title: fields.title,
    description: fields.description,
    points: fields.points,
    dueDate: fields.dueDate,
    recurrence: fields.recurrence,
    recurrenceDays: fields.recurrenceDays,
    assignedToUserId: fields.assignedToUserId,
  });

  revalidatePath("/chores");
  revalidatePath("/");
  revalidatePath("/calendar");
  return null;
}

export async function updateChore(
  _prevState: FormState,
  formData: FormData,
): Promise<FormState> {
  const { household, user } = await requireActiveUser();
  requireAdmin(user);

  const id = String(formData.get("id") ?? "");
  const existing = await db.query.chores.findFirst({
    where: and(eq(chores.id, id), eq(chores.householdId, household.id)),
  });
  if (!existing) return { error: "Chore not found." };

  const fields = readChoreFields(formData);
  if ("error" in fields) return { error: fields.error };

  const assignee = await db.query.users.findFirst({
    where: and(
      eq(users.id, fields.assignedToUserId),
      eq(users.householdId, household.id),
    ),
  });
  if (!assignee) return { error: "Assignee not found." };

  await db
    .update(chores)
    .set({
      title: fields.title,
      description: fields.description,
      points: fields.points,
      dueDate: fields.dueDate,
      recurrence: fields.recurrence,
      recurrenceDays: fields.recurrenceDays,
      assignedToUserId: fields.assignedToUserId,
    })
    .where(eq(chores.id, id));

  revalidatePath("/chores");
  revalidatePath("/");
  revalidatePath("/calendar");
  return null;
}

export async function deleteChore(formData: FormData) {
  const { household, user } = await requireActiveUser();
  requireAdmin(user);

  const id = String(formData.get("id") ?? "");
  await db
    .delete(chores)
    .where(and(eq(chores.id, id), eq(chores.householdId, household.id)));

  revalidatePath("/chores");
  revalidatePath("/");
  revalidatePath("/calendar");
}

function todayRange() {
  const start = new Date();
  start.setHours(0, 0, 0, 0);
  const end = new Date(start);
  end.setDate(end.getDate() + 1);
  return { start, end };
}

export async function completeChore(formData: FormData) {
  const { household, user } = await requireActiveUser();

  const choreId = String(formData.get("choreId") ?? "");
  const chore = await db.query.chores.findFirst({
    where: and(eq(chores.id, choreId), eq(chores.householdId, household.id)),
  });
  if (!chore) return;

  // Anyone can mark their own chore done; admins can mark on behalf of
  // someone else too (e.g. a parent checking off a younger kid's chore).
  if (chore.assignedToUserId !== user.id && user.role !== "admin") {
    throw new Error("This chore isn't assigned to you.");
  }
  if (!chore.assignedToUserId) return;

  const { start, end } = todayRange();
  const alreadyDoneToday = await db.query.choreCompletions.findFirst({
    where: and(
      eq(choreCompletions.choreId, chore.id),
      gte(choreCompletions.completedAt, start),
      lt(choreCompletions.completedAt, end),
    ),
  });
  if (alreadyDoneToday) return;

  const assignedToUserId = chore.assignedToUserId;

  db.transaction((tx) => {
    tx.insert(choreCompletions)
      .values({
        id: crypto.randomUUID(),
        choreId: chore.id,
        userId: assignedToUserId,
        pointsAwarded: chore.points,
      })
      .run();

    tx.update(users)
      .set({ points: sql`${users.points} + ${chore.points}` })
      .where(eq(users.id, assignedToUserId))
      .run();
  });

  revalidatePath("/chores");
  revalidatePath("/");
  revalidatePath("/calendar");
}

export async function uncompleteChore(formData: FormData) {
  const { household, user } = await requireActiveUser();

  const choreId = String(formData.get("choreId") ?? "");
  const chore = await db.query.chores.findFirst({
    where: and(eq(chores.id, choreId), eq(chores.householdId, household.id)),
  });
  if (!chore || !chore.assignedToUserId) return;

  if (chore.assignedToUserId !== user.id && user.role !== "admin") {
    throw new Error("This chore isn't assigned to you.");
  }

  const { start, end } = todayRange();
  const completion = await db.query.choreCompletions.findFirst({
    where: and(
      eq(choreCompletions.choreId, chore.id),
      gte(choreCompletions.completedAt, start),
      lt(choreCompletions.completedAt, end),
    ),
  });
  if (!completion) return;

  const assignedToUserId = chore.assignedToUserId;

  db.transaction((tx) => {
    tx.delete(choreCompletions)
      .where(eq(choreCompletions.id, completion.id))
      .run();

    tx.update(users)
      .set({ points: sql`${users.points} - ${completion.pointsAwarded}` })
      .where(eq(users.id, assignedToUserId))
      .run();
  });

  revalidatePath("/chores");
  revalidatePath("/");
  revalidatePath("/calendar");
}

File: backend/internal/api/handlers/chores.go
package handlers

import (
	"errors"
	"net/http"
	"time"

	"github.com/gin-gonic/gin"

	"hestia/backend/internal/api/middleware"
	"hestia/backend/internal/services"
)

var validRecurrences = map[string]bool{
	"none": true, "daily": true, "weekly": true, "weekdays": true, "custom": true,
}

func (d *Deps) ListChores(c *gin.Context) {
	household := middleware.CurrentHousehold(c)
	dueToday := c.Query("due") == "today"

	chores, err := d.Chore.List(household.ID, dueToday)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to list chores"})
		return
	}
	c.JSON(http.StatusOK, gin.H{"chores": chores})
}

func (d *Deps) GetChore(c *gin.Context) {
	household := middleware.CurrentHousehold(c)
	chore, err := d.Chore.Get(household.ID, c.Param("id"))
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "chore not found"})
		return
	}
	c.JSON(http.StatusOK, chore)
}

type choreRequest struct {
	Title            string  `json:"title"`
	Description      *string `json:"description"`
	Points           int     `json:"points"`
	AssignedToUserID string  `json:"assignedToUserId"`
	Recurrence       string  `json:"recurrence"`
	DueDate          string  `json:"dueDate"` // YYYY-MM-DD
	RecurrenceDays   []int   `json:"recurrenceDays"`
}

func (r *choreRequest) toInput() (services.ChoreInput, string) {
	if r.Title == "" {
		return services.ChoreInput{}, "title is required"
	}
	if r.AssignedToUserID == "" {
		return services.ChoreInput{}, "choose who this is assigned to"
	}
	if r.Points < 0 {
		return services.ChoreInput{}, "points must be zero or a positive number"
	}
	if r.Recurrence == "" {
		r.Recurrence = "none"
	}
	if !validRecurrences[r.Recurrence] {
		return services.ChoreInput{}, "invalid recurrence"
	}
	dueDate, err := time.ParseInLocation("2006-01-02", r.DueDate, time.Local)
	if err != nil {
		return services.ChoreInput{}, "choose a valid due date"
	}
	var recurrenceDays *string
	if r.Recurrence == "custom" {
		if len(r.RecurrenceDays) == 0 {
			return services.ChoreInput{}, "pick at least one day for a custom schedule"
		}
		s := services.SerializeRecurrenceDays(r.RecurrenceDays)
		recurrenceDays = &s
	}

	return services.ChoreInput{
		Title:            r.Title,
		Description:      r.Description,
		Points:           r.Points,
		AssignedToUserID: r.AssignedToUserID,
		Recurrence:       r.Recurrence,
		DueDate:          dueDate,
		RecurrenceDays:   recurrenceDays,
	}, ""
}

func (d *Deps) CreateChore(c *gin.Context) {
	household := middleware.CurrentHousehold(c)

	var req choreRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid request"})
		return
	}
	input, errMsg := req.toInput()
	if errMsg != "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": errMsg})
		return
	}

	chore, err := d.Chore.Create(household.ID, input)
	if err != nil {
		if errors.Is(err, services.ErrNotFound) {
			c.JSON(http.StatusBadRequest, gin.H{"error": "assignee not found"})
			return
		}
		c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to create chore"})
		return
	}
	c.JSON(http.StatusCreated, chore)
}

func (d *Deps) UpdateChore(c *gin.Context) {
	household := middleware.CurrentHousehold(c)

	var req choreRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid request"})
		return
	}
	input, errMsg := req.toInput()
	if errMsg != "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": errMsg})
		return
	}

	chore, err := d.Chore.Update(household.ID, c.Param("id"), input)
	if err != nil {
		if errors.Is(err, services.ErrNotFound) {
			c.JSON(http.StatusNotFound, gin.H{"error": "chore not found"})
			return
		}
		c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to update chore"})
		return
	}
	c.JSON(http.StatusOK, chore)
}

func (d *Deps) DeleteChore(c *gin.Context) {
	household := middleware.CurrentHousehold(c)
	if err := d.Chore.Delete(household.ID, c.Param("id")); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to delete chore"})
		return
	}
	c.JSON(http.StatusOK, gin.H{"ok": true})
}

func (d *Deps) CompleteChore(c *gin.Context) {
	household := middleware.CurrentHousehold(c)
	user := middleware.CurrentUser(c)
	choreID := c.Param("id")

	completion, err := d.Chore.Complete(household.ID, choreID, user.ID, user.Role)
	if err != nil {
		switch {
		case errors.Is(err, services.ErrNotFound):
			c.JSON(http.StatusNotFound, gin.H{"error": "chore not found"})
		case errors.Is(err, services.ErrForbidden):
			c.JSON(http.StatusForbidden, gin.H{"error": "this chore isn't assigned to you"})
		case errors.Is(err, services.ErrUnassigned):
			c.JSON(http.StatusBadRequest, gin.H{"error": "chore has no assignee"})
		case errors.Is(err, services.ErrAlreadyDone):
			c.JSON(http.StatusOK, gin.H{"ok": true, "alreadyDone": true})
		default:
			c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to complete chore"})
		}
		return
	}

	d.broadcastEvent("chore.completed", map[string]any{"choreId": choreID, "userId": completion.UserID})
	c.JSON(http.StatusOK, completion)
}

func (d *Deps) UncompleteChore(c *gin.Context) {
	household := middleware.CurrentHousehold(c)
	user := middleware.CurrentUser(c)
	choreID := c.Param("id")

	err := d.Chore.Uncomplete(household.ID, choreID, user.ID, user.Role)
	if err != nil {
		switch {
		case errors.Is(err, services.ErrNotFound):
			c.JSON(http.StatusNotFound, gin.H{"error": "chore not found"})
		case errors.Is(err, services.ErrForbidden):
			c.JSON(http.StatusForbidden, gin.H{"error": "this chore isn't assigned to you"})
		case errors.Is(err, services.ErrUnassigned), errors.Is(err, services.ErrNotCompleted):
			c.JSON(http.StatusOK, gin.H{"ok": true})
		default:
			c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to uncomplete chore"})
		}
		return
	}

	d.broadcastEvent("chore.uncompleted", map[string]any{"choreId": choreID, "userId": user.ID})
	c.JSON(http.StatusOK, gin.H{"ok": true})
}
